### DISCUSSION TITLE
a script that used to work in previous versions doesn't work in (1.11.x)

### DISCUSSION THREAD

**Discussion by Bricklam** [freeplane/freeplane#1217] [Help]
I used this script in the past to copy & paste node or branch to all the selected nodes, but now it doesn't work: I don't know why

```
import org.freeplane.features.map.mindmapmode.clipboard.MMapClipboardController

def clipboardController = MMapClipboardController.controller
def target = node.delegate
clipboardController.paste(clipboardController.clipboardContents, target, false, target.isNewChildLeft())
```

I get this error message:
"No signature of method: org.freeplane.features.map.NodeModel.isNewChildLeft() is applicable for argument types: () values: [] at line -1"

Thanks

**Comment by macmarrum** [comment]
Freeplane's internal API changed in 1.11.x.
It's better to use the Scripting API for that reason.
Here's an appropriate script: [pasteToSelected.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/pasteToSelected.groovy)

Note: Remove or change `="/menu_bar/Mac1/Paste"` in the first line to have the script appear in a different menu location

**Comment by Bricklam** [comment]
Thanks @macmarrum, your script works fine.
 but I want to implement the using Freeplane API, instead of just calling the paste menu action.

I'm aware that Freeplane's internal API has changed in 1.11.x, and I have checked the source code file:
`"\1.11.5_pre05\freeplane-1.11.5-pre05\freeplane\src\main\java\org\freeplane\features\map\mindmapmode\clipboard\MMapClipboardController.java"`
 in order to know how should I change the original script I posted above, in order to be compatible with the new API changes, but unfortunately couldn't find a way.

can you help me find a way to change the original script according to the new API to make it work again?

appreciate your help
Thanks

**Reply by macmarrum** [reply]
> I want to implement the using Freeplane API, instead of just calling the paste menu action.

That's fine with me.
I am interested in learning your reasons to avoid calling the paste menu action.

**Comment by Bricklam** [comment]
I think that I have found a simple solution, but I'm not sure this is the correct way to do it:

just by removing two arguments from this line:  (removing the part "false, target.isNewChildLeft()"
`// clipboardController.paste(clipboardController.clipboardContents, target, false, target.isNewChildLeft())`

to be like this:
`clipboardController.paste(clipboardController.clipboardContents, target)`

and it now works

Thanks

**Reply by macmarrum** [reply]
It looks alright to me :+1:

**Comment by bekoeppel** [comment]
I had the same question, and am also using the internal API instead of the menu actions. As you asked:
> I am interested in learning your reasons to avoid calling the paste menu action.

What I'm doing is adding a new menu action, to _paste as sibling_. I would like to paste something in my clipboard on top of a list of nodes. Without my script, I have to select the parent node, paste, and then move whatever I pasted manually to the top. With my script, I can select the first node on a list, and run PasteAsSiblingBefore.

In a sense, it's similar to the Insert New Previous Sibling Node action that comes out of the box. Might be useful to have the same functionality built-in for pasting.


```
import org.freeplane.features.map.mindmapmode.clipboard.MMapClipboardController
import org.freeplane.core.ui.components.UITools
import org.freeplane.features.map.NodeModel.Side;

/**
 * Paste clipboard as sibling, before the current node. Useful to paste stuff
 * at the top of a list. Won't work on root node.
 */

def clipboardController = MMapClipboardController.controller
def target = node.delegate

// check for root node
if (map.getRootNode().equals(target) || target.getParentNode() == null) {
    UITools.errorMessage("Can't paste as sibling on root node");
    return;
}

clipboardController.paste(clipboardController.clipboardContents, target, Side.AS_SIBLING)
```

**Reply by euu2021** [reply]
This is interesting. I think there are 7 main maneuvers that could be useful when pasting or creating a new node:

![image](https://github.com/freeplane/freeplane/assets/77707706/2ccb718b-872f-4e5f-9d97-53ff6cffa949)

It would be especially useful if there was a hotkeys combination that make those maneuvers intuitive. Right now, what feels intuitive to me  is  Ctrl for the down maneuvers, and Shift for the up maneuvers. Perhaps the Alt can have the meaning of "in the children level", as it is more to the right than Ctrl and Shift, in the keyboard.

Related discussion (there, about drag and drop): #1357

**Reply by walmar** [reply]
@bekoeppel, thank you for sharing the script! It does exactly what I was looking for.

Although I have to point out that with the *drag and drop as sibling before or after* feature (https://github.com/freeplane/freeplane/commit/a9ade60eb2860c930852ffcf71d7746d06685cfa) in latest Freeplane 1.12.10, the constant ``AS_SIBLING`` has been changed to ``AS_SIBLING_BEFORE`` and ``AS_SIBLING_AFTER`` (by the way, to paste as child, it seems to be [DEFAULT](https://docs.freeplane.org/api/org/freeplane/api/Side.html#DEFAULT)).
So I have a script for each of them and have assigned them hotkeys that I send via middle mouse button gestures (Ctrl + drag up or down) from an AutoHotKey script, like I've been doing for other actions for a few years now.

I also find out that pasting as sibling node was already [requested back in 2010 for FreeMind](https://sourceforge.net/p/freemind/feature-requests/776/).

Personally, I would prefer to use drag-and-drop, but that's usually not yet possible across maps or without copying children, as the related feature requests (#2347, #2359, #2372) from the last weeks to solve the situation are still open. Of course, drag-and-drop generally remains the second best method after cut/copy and paste, as it is less flexible, especially if someone wants to paste from other programs.
